### ISSUE TITLE
Join by dictionary

### ISSUE THREAD

**Issue by Joozty** [yandex/ClickHouse#2612]
Hi guys, 

is it possible to JOIN BY DICTIONARY. I can see only [ARRAY JOIN clause](https://clickhouse.yandex/docs/en/query_language/queries/#array-join-clause) or the only way how to achieve this is to create two separate tables and than create JOIN query.

The idea is following

I've got this record

```
ID | YEAR_DICT
1  | {2015: 10, 2016: 15, 2017: 5}
```

and I want this output: 

```
ID | YEAR | NUMBER
1  | 2015 | 10
1  | 2016 | 15
1  | 2017 | 5
```

Thanks

**Comment by filimonov** [comment]
What do you mean by DICTIONARY? JSON field?

**Comment by Joozty** [comment]
@filimonov I already solved it with `ARRAY JOIN` clause. Instead of using JSON dict I decided to use array like this: `[[key, value], [key, value] ...]`

It works well. The only problem I have now is with MATERIALIZED VIEW. (a similar question was posted on [Stackoverflow](https://stackoverflow.com/questions/51233488/update-materialized-view-with-join-statement)) I have exactly the same tables as in that question and when I create materialized view it is always empty. Even when I insert new row to table `A` or `B`.

Do you have some workaround for this problem?

Thanks

**Comment by filimonov** [comment]
> @filimonov I already solved it with ARRAY JOIN clause. Instead of using JSON dict I decided to use array like this: [[key, value], [key, value] ...]

Also array of tuples also could be a nice solution.
Like 
```
select dict.1 as key, dict.2 as value from system.one array join [ tuple('key', 'value'), tuple('key', 'value') ] as dict;
```

> It works well. The only problem I have now is with MATERIALIZED VIEW. (a similar question was posted on Stackoverflow) I have exactly the same tables as in that question and when I create materialized view it is always empty. Even when I insert new row to table A or B.

There is no way to create a materialized view from 2 tables. Materialized views work more like 'on insert' triggers. You can't make one insert trigger for 2 tables...

**Comment by Joozty** [comment]
@filimonov Is it possible to refresh a materialized view manually at least?

**Comment by filimonov** [comment]
1) very often table can be redesigned to avoid JOINs in Clickhouse (using Arrays / external Dictionaries or denormalization).
2) you can just have a table and put data there on some regular basis like that:
```INSERT INTO ... SELECT ... WHERE ... ```

**Comment by Joozty** [comment]
I tried to use arrays to avoid `JOIN` but I wasn't successful because there is no possibility to use `UNION ALL` in materialized views. And I needed it. 

To show you what I am trying to achieve. 

The table contains `ID, one_array, second_array`.  These arrays are in following format `[[int, int], [int, int]...]`. Every inside array has two elements. First is a year and second is some number. 

```
ID | one_array                  | second_array
------------------------------------------------------------
1  | [[2016, 123], [2018, 456]] | [[2015, 753], [2018, 963]]
```

I want to achieve this ouput:
```
ID | year | one_array_num | second_array_num
1  | 2015 | null          | 753
1  | 2016 | 123           | null
1  | 2018 | 456           | 963
```
 But as I mentioned above. I cannot use `UNION ALL`. So I decided to use two tables - first one contains `id` and some other stuff,  second one contains `id`, `year`, `one_array_num`,  `second_array_num`). And these two tables I am trying to join in materialized view. 

The table is very big and I don't want to run join every time when some query executes.
